lib/widgets/compose_box.dart

```diff
@@ -362,6 +363,32 @@ class _ContentInputState extends State<_ContentInput> with WidgetsBindingObserve
     }
   }
 
+  void _handleContentInserted(KeyboardInsertedContent content) async {
+    if (!content.hasData) {
+      // The data can be empty when the URL is associated with an empty
+      // resource.  See Flutter engine implementation that provides this data:
+      //   https://github.com/flutter/flutter/blob/0ffc4ce00ea7bb912e379adf39354644eab2c17e/engine/src/flutter/shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java#L497-L548
```

Comment:
This comment is specifically about the case where `data` ends up being present as an empty array, right? In the code you linked to, I assume that specifically means `readStreamFully` came up empty in line 531:

https://github.com/flutter/flutter/blob/0ffc4ce00/engine/src/flutter/shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java#L531
```java
final byte[] data = this.readStreamFully(is, 64 * 1024);
```

(Right?)

How about a different case: what could cause `data` to be null as opposed to present-but-empty? I noticed [`KeyboardInsertedContent.data`](https://api.flutter.dev/flutter/services/KeyboardInsertedContent/data.html) is optional. This would be helpful to document upstream there. 🙂

Also, looking around the whole `commitContent` function you linked to, it looks like there are various unhappy paths there, where the function can't do something desirable and returns `false`. Like if the Android version is too old. What's the user-facing behavior in those unhappy paths?

Reply:
Having swept through the Java code, I couldn't find other places where `data` is `null` and gets passed via a `textInputChannel.commitContent` call. The link points to the only current non-test call site. The Dart code constructor doesn't have code paths where `data` is explicitly `null` either.

`data` can be empty if (a) the file is empty or (b) there was an `IOException` during the read. So `data` should probably be `null` if there is an error reading it, so that we can handle it properly.

---

When returning `false`, no `textInputChannel.commitContent` call is made and our handler does not get called at all. But the return value is probably read by some part of the Android SDK code: https://developer.android.com/reference/android/view/inputmethod/BaseInputConnection

> Default implementation which invokes [View.performReceiveContent](https://developer.android.com/reference/android/view/View#performReceiveContent(android.view.ContentInfo)) on the target view if the view [allows](https://developer.android.com/reference/android/view/View#getReceiveContentMimeTypes()) content insertion; otherwise returns false without any side effects.

> true if this request is accepted by the application, whether the request is already handled or still being handled in background, false otherwise.

I assume the behavior is akin to when we do not provide `ContentInsertionConfiguration` at all. That also results in this method returning `false` because the `allowedMimeTypes` default to an empty list. This gives you an OS-specific error message:
![image](https://github.com/user-attachments/assets/93afeeea-fdde-462e-b86d-18cf621124d9)